### PR TITLE
add first attempt at file load/save

### PR DESCRIPTION
**Pull request by drscotthawley**
Addresses https://github.com/keunwoochoi/torchaudio-contrib/issues/31

### CODE REVIEW

File: torchaudio-contrib/io.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    if scipy_ok:
+        if mono and (len(signal.shape) > 1):     # convert to mono (by truncating other channels)
+            signal = signal[:,0]
+
+        if isinstance(signal[0], np.int16):      # convert from ints to floats if necessary
+            signal = np.array(signal, dtype=dtype)/32767.0   # change from [-32767..32767] to [-1..1]
+
+        if out_sr != int(sr):
+            print(f"read_audio_file: Got sample rate of {rate} Hz instead of {sr} Hz requested. Resampling.")
+            signal = librosa.resample(signal, rate*1.0, sr*1.0, res_type='kaiser_fast')
```
Comment (deNsuh) [inline torchaudio-contrib/io.py]: Is there a reason that you're converting these into `float`s? It doesn't seem to matter looking at the [official docs](https://librosa.github.io/librosa/generated/librosa.core.resample.html#librosa.core.resample).
```suggestion
            signal = librosa.resample(signal, rate, sr, res_type='kaiser_fast')
```

File: torchaudio-contrib/io.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
+import torch
+import numpy as np
+import warnings
+from scipy.io import wavfile
+import librosa
```
Comment (f0k) [inline torchaudio-contrib/io.py]: This means users need to install librosa even if they only read wav files. The import should only be done when needed.

File: torchaudio-contrib/io.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    """
+    # first try to read via scipy, because it's fast
+    scipy_ok = False
+    with warnings.catch_warnings(record=True) as w:
+        warnings.simplefilter("error")    # scipy throws warnings which should be errors
+        try:
+            out_sr, signal = wavfile.read(filename)
+            scipy_ok = True
+        except wavfile.WavFileWarning:
+            print("read_audio_file: Warning raised by scipy. ",end="")
```
Comment (f0k) [inline torchaudio-contrib/io.py]: As long as torch supports Python 2, we should too. Would need a `from __future__ import print_function`, or a way that does not need the `end` keyword argument. I mean, you could just print everything when you reach librosa.

File: torchaudio-contrib/io.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+    if scipy_ok:
+        if mono and (len(signal.shape) > 1):     # convert to mono (by truncating other channels)
+            signal = signal[:,0]
+
+        if isinstance(signal[0], np.int16):      # convert from ints to floats if necessary
```
Comment (f0k) [inline torchaudio-contrib/io.py]: Firstly, this should be optional, and secondly, it's missing 8-bit and 24-bit conversion.

File: torchaudio-contrib/io.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+            signal = np.array(signal, dtype=dtype)/32767.0   # change from [-32767..32767] to [-1..1]
```
Comment (f0k) [inline torchaudio-contrib/io.py]: This creates a copy of the full signal as the given dtype, even if the given dtype is `int16`, and then creates another copy of the full signal with the scaled values. Should be done in a single step with `np.multiply`.
Note that if you pass `np.int16` as the dtype, you will get a `float64` result (that will be converted back to int16 in the end, but takes up 4 times the space in the meantime).

File: torchaudio-contrib/io.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    scipy_ok = False
+    with warnings.catch_warnings(record=True) as w:
+        warnings.simplefilter("error")    # scipy throws warnings which should be errors
+        try:
+            out_sr, signal = wavfile.read(filename)
+            scipy_ok = True
+        except wavfile.WavFileWarning:
+            print("read_audio_file: Warning raised by scipy. ",end="")
+
+    if scipy_ok:
```
Comment (f0k) [inline torchaudio-contrib/io.py]: That's not a good style. I understand you're doing this to get out of the `with` block, but a better solution would be wrapping the "warnings-as-exceptions" scipy call in a separate function, or can't you just move it inside `try:`?

File: torchaudio-contrib/io.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        if out_sr != int(sr):
+            print(f"read_audio_file: Got sample rate of {rate} Hz instead of {sr} Hz requested. Resampling.")
+            signal = librosa.resample(signal, rate*1.0, sr*1.0, res_type='kaiser_fast')
+
+    else:                                         # try librosa; it's slower but general
+        print("Trying librosa.")
+        signal, out_sr = librosa.core.load(filename, mono=mono, sr=sr, res_type='kaiser_fast')
+
+    if norm:
+        signal = signal/np.max(np.abs(signal))
```
Comment (f0k) [inline torchaudio-contrib/io.py]: Again an unneeded copy of the full signal. Should be inplace.

File: torchaudio-contrib/io.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+    return torch.from_numpy(signal.astype(dtype)).to(device), out_sr
```
Comment (f0k) [inline torchaudio-contrib/io.py]: And again an extra copy of the signal (`astype` always copies).

**Comment by hagenw** [comment]
Just a general comment: I would be very much in favor of adding directly corresponding tests for `load` and `save`.

**Comment by f0k** [comment]
I'm going to close this, since we haven't decided on the API yet (#31), but it's helpful to have such an attempt to spur the discussion and we may want to reuse parts of it later!

File: torchaudio-contrib/io.py (hunk 1 above)
Comment (drscotthawley) [reply torchaudio-contrib/io.py]: For some reason I thought librosa expected floats for those variables. That was wrong.

File: torchaudio-contrib/io.py (hunk 2 above)
Comment (drscotthawley) [reply torchaudio-contrib/io.py]: - The librosa import is to handle WAV file types which scipy cannot -- e.g. 24 bit files.

File: torchaudio-contrib/io.py (hunk 4 above)
Comment (drscotthawley) [reply torchaudio-contrib/io.py]: - The 8-bit and 24-bit cases will trigger warnings from scipy and will kick the routine into librosa.

File: torchaudio-contrib/io.py (hunk 5 above)
Comment (drscotthawley) [reply torchaudio-contrib/io.py]: Thanks for catching this.   I was unaware that astype copied by default;  one can easily add `copy=False`, but yea, that dtype should go on the outside, not the inside!

File: torchaudio-contrib/io.py (hunk 6 above)
Comment (drscotthawley) [reply torchaudio-contrib/io.py]: This is *recommended* style [straight from the Python Docs](https://docs.python.org/3/library/warnings.html#testing-warnings).

File: torchaudio-contrib/io.py (hunk 7 above)
Comment (drscotthawley) [reply torchaudio-contrib/io.py]: Thanks for catching that.

File: torchaudio-contrib/io.py (hunk 3 above)
Comment (drscotthawley) [reply torchaudio-contrib/io.py]: Good point. I haven't thought about Python 2 in a couple years.

File: torchaudio-contrib/io.py (hunk 2 above)
Comment (f0k) [reply torchaudio-contrib/io.py]: Okay, let me refine: This means users need to isntall librosa even if they only read 16-bit wav files. The import should only be done when needed.

File: torchaudio-contrib/io.py (hunk 6 above)
Comment (f0k) [reply torchaudio-contrib/io.py]: I meant the technique of setting a boolean flag somewhere and then checking for it later.

File: torchaudio-contrib/io.py (hunk 5 above)
Comment (f0k) [reply torchaudio-contrib/io.py]: If the dtype goes on the outside (I assume you meant `np.array(signal / magic_number, dtype=dtype)`), you're still creating a temporary float64 copy that's converted afterwards. I'd suggest doing the division and conversion in one step using `np.multiply`.

**Pull request closed without merging**